File: src/forward_model/builtins/surface.zig
const std = @import("std");
const Scene = @import("../../input/Scene.zig").Scene;
const PreparedOpticalState = @import("../optical_properties/root.zig").PreparedOpticalState;
const ForwardResult = @import("../radiative_transfer/root.zig").ForwardResult;

pub const EvaluationContext = struct {
    scene: *const Scene,
    prepared: *const PreparedOpticalState,
    wavelength_nm: f64,
    safe_span: f64,
    phase: f64,
    forward: ForwardResult,
};

pub const Provider = struct {
    id: []const u8,
    brdfFactor: *const fn (context: EvaluationContext) f64,
};

pub fn resolve(provider_id: []const u8) ?Provider {
    if (std.mem.eql(u8, provider_id, "builtin.lambertian_surface")) {
        return .{
            .id = provider_id,
            .brdfFactor = lambertianBrdfFactor,
        };
    }
    return null;
}

fn lambertianBrdfFactor(context: EvaluationContext) f64 {
    _ = context.prepared;
    _ = context.wavelength_nm;
    _ = context.safe_span;
    _ = context.phase;
    _ = context.forward;
    return switch (context.scene.surface.kind) {
        // Lambertian BRDF is isotropic, so the directional factor stays unity here.
        // Scene albedo already enters the radiative transfer routine through ForwardInput.surface_albedo.
        .lambertian => 1.0,
        // Wavelength-dependent surfaces use the same isotropic directional factor;
        // spectral dependence is handled through the albedo schedule.
        .wavel_dependent => 1.0,
    };
}

File: rust_src/forward_model/implementations/surface.rs
use crate::{
    forward_model::{
        optical_properties::state_build::PreparedOpticalState,
        radiative_transfer::common_types::ForwardResult,
    },
    input::{scene::Scene, surface::Kind},
};

pub const LAMBERTIAN_SURFACE_ID: &str = "builtin.lambertian_surface";

pub struct EvaluationContext<'a> {
    pub scene: &'a Scene,
    pub prepared: &'a PreparedOpticalState,
    pub wavelength_nm: f64,
    pub safe_span: f64,
    pub phase: f64,
    pub forward: &'a ForwardResult,
}

pub type BrdfFactorFn = for<'a> fn(EvaluationContext<'a>) -> f64;

#[derive(Debug, Clone, Copy)]
pub struct Implementation {
    pub id: &'static str,
    pub brdf_factor: BrdfFactorFn,
}

pub fn resolve(provider_id: &str) -> Option<Implementation> {
    if provider_id == LAMBERTIAN_SURFACE_ID {
        return Some(Implementation {
            id: LAMBERTIAN_SURFACE_ID,
            brdf_factor: lambertian_brdf_factor,
        });
    }
    None
}

fn lambertian_brdf_factor(context: EvaluationContext<'_>) -> f64 {
    let _ = context.prepared;
    let _ = context.wavelength_nm;
    let _ = context.safe_span;
    let _ = context.phase;
    let _ = context.forward;
    match context.scene.surface.kind {
        // Albedo already enters the RTM input, so isotropic surfaces keep this directional factor at one.
        Kind::Lambertian | Kind::WavelDependent => 1.0,
    }
}
